src/osd/PG.cc

```diff
@@ -4517,17 +4520,24 @@ void PG::_repair_oinfo_oid(ScrubMap &smap)
     } catch(...) {
       continue;
     }
-    if (oi.soid != hoid) {
+    if (oi.soid != hoid || (skip_data_digest && oi.is_data_digest())) {
       ObjectStore::Transaction t;
       OSDriver::OSTransaction _t(osdriver.get_transaction(&t));
+      ostringstream errormsg;
+      // Fix object info
+      if (oi.soid != hoid) {
+        errormsg << " bad oid in object info: " << oi.soid;
+        oi.soid = hoid;
+      }
+      if (skip_data_digest && oi.is_data_digest()) {
+        errormsg << " data_digest should not be set";
+        oi.clear_data_digest();
+      }
       osd->clog->error() << "osd." << osd->whoami
 			    << " found object info error on pg "
-			    << info.pgid
-			    << " oid " << hoid << " oid in object info: "
-			    << oi.soid
+			    << info.pgid << " oid " << hoid
+			    << errormsg.str()
 			    << "...repaired";
-      // Fix object info
-      oi.soid = hoid;
```

Comment:
I think we *just* want this chunk here (and up), at least for this patch.  The changes I have in #23051 address the other issues with a new osd_distrust_data_digest option.

Reply:
The semi-related issue is that the skip_data_digest logic is weak because it depends on whether bluestore is used on the primary, and if so, it clears the digest for all replicas.  But other replicas might be filestore.  Do we want to skip digests in that case?  Probably not?

Probably not worth fixing that right now.  But this change kind of amplifies that choice in that we are throwing out checksums on all replicas on account of the primary...

How hard would it be to make deep scrub throw out the checksum *only* if we have skip_data_digests (or distrust_data_digests) and the replicas match but the oi digest doesn't?  That's a more targetted fix that doesn't generate a repair operation on basically every coldish object in the system..

Reply:
@liewegas Actually, this code is run independently on each replica.  I hadn't thought about this, but the data_digest would NOT be cleared on a replica that is using filestore.